File: src/download_data.py
import requests
import gzip
import shutil
from pathlib import Path
import config
import zipfile
import io

def download_file(url, dest_path):
    if dest_path.exists():
        print(f"File already exists: {dest_path}")
        return

    print(f"Downloading {url}...")
    try:
        r = requests.get(url, stream=True)
        r.raise_for_status()
        with open(dest_path, 'wb') as f:
            for chunk in r.iter_content(chunk_size=8192):
                f.write(chunk)
        print(f"Downloaded to {dest_path}")
    except Exception as e:
        print(f"Failed to download {url}: {e}")

def main():
    # 1. Download LODES WAC Data (Employment)
    for year in config.YEARS:
        filename = f"{config.STATE}_wac_S000_JT00_{year}.csv.gz"
        url = f"{config.LODES_BASE_URL}/{config.STATE}/wac/{filename}"
        dest = config.DATA_RAW / filename
        download_file(url, dest)

    # 2. Download Geography (Block Groups Shapefile)
    # Using TIGER 2020 Block Groups for WA to match LODES 8 (2020 Geography)
    geo_filename = f"tl_2020_{config.STATE}_bg.zip"
    # FIPS code for WA is 53
    geo_url = "https://www2.census.gov/geo/tiger/TIGER2020/BG/tl_2020_53_bg.zip"
    geo_dest = config.DATA_RAW / geo_filename
    download_file(geo_url, geo_dest)
    
    # Extract Shapefile
    extract_dir = config.DATA_RAW / "shapefiles"
    if not extract_dir.exists():
        print("Extracting shapefile...")
        try:
            with zipfile.ZipFile(geo_dest, 'r') as zip_ref:
                zip_ref.extractall(extract_dir)
        except Exception as e:
            print(f"Error extracting shapefile: {e}")

if __name__ == "__main__":
    main()

File: src/download_data.R
# Download Data Script
source("src/config.R")

download_data <- function() {
  # 1. Download LODES WAC Data (Employment)
  message("--- Downloading LODES Data ---")
  for (year in YEARS) {
    filename <- sprintf("%s_wac_S000_JT00_%d.csv.gz", STATE, year)
    url <- sprintf("%s/%s/wac/%s", LODES_BASE_URL, STATE, filename)
    dest_path <- file.path(DATA_RAW, filename)
    
    if (file.exists(dest_path)) {
      message(sprintf("File already exists: %s", filename))
    } else {
      message(sprintf("Downloading %s...", url))
      tryCatch({
        download.file(url, dest_path, mode = "wb")
        message(sprintf("Downloaded to %s", dest_path))
      }, error = function(e) {
        message(sprintf("Failed to download %s: %s", url, e$message))
      })
    }
  }
  
  # 2. Download Geography (Block Groups Shapefile)
  # Using TIGER 2020 Block Groups for WA
  message("\n--- Downloading Geography Data ---")
  geo_filename <- sprintf("tl_2020_%s_bg.zip", STATE)
  # FIPS code for WA is 53
  geo_url <- "https://www2.census.gov/geo/tiger/TIGER2020/BG/tl_2020_53_bg.zip"
  geo_dest <- file.path(DATA_RAW, geo_filename)
  
  if (file.exists(geo_dest)) {
    message(sprintf("File already exists: %s", geo_filename))
  } else {
    message(sprintf("Downloading %s...", geo_url))
    tryCatch({
      download.file(geo_url, geo_dest, mode = "wb")
      message(sprintf("Downloaded to %s", geo_dest))
    }, error = function(e) {
      message(sprintf("Failed to download %s: %s", geo_url, e$message))
    })
  }
  
  # Extract Shapefile
  extract_dir <- file.path(DATA_RAW, "shapefiles")
  if (!dir.exists(extract_dir)) {
    message("Extracting shapefile...")
    tryCatch({
      unzip(geo_dest, exdir = extract_dir)
      message("Extraction complete.")
    }, error = function(e) {
      message(sprintf("Error extracting shapefile: %s", e$message))
    })
  } else {
    message("Shapefiles directory already exists.")
  }
}

if (sys.nframe() == 0) {
  download_data()
}
